### ISSUE TITLE
Invalid device path for SSD on Dell SLA

### ISSUE THREAD

**Issue by jiazhang0** [vathpela/fwup#15]
The generated fwupdate status variable contains an invalid device path on my DELL SLA board. On this board, an USB key (/dev/sdb) and SSD(/dev/sda) are attached. Two drives all uses the first partition as ESP (MBR scheme not GPT). See the infos below.

```
# fdisk -l /dev/sda
Disk /dev/sda: 29.8 GiB, 32017047552 bytes, 62533296 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x5078fe0f

Device    Boot Start       End Blocks  Id System
/dev/sda1         63    273104 136521   e W95 FAT16 (LBA)

# fdisk -l /dev/sdb
Disk /dev/sdb: 7.5 GiB, 8004304896 bytes, 15633408 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x59c88c89

Device    Boot     Start       End  Blocks  Id System
/dev/sdb1           2048    264191  131072   c W95 FAT32 (LBA)
/dev/sdb2         264192   6555647 3145728  83 Linux
/dev/sdb3        6555648  12847103 3145728  83 Linux
/dev/sdb4       12847104  15633407 1393152  83 Linux

# mount | grep sd
/dev/sdb2 on / type ext3 (rw,relatime,errors=continue,user_xattr,barrier=1,data=ordered)
/dev/sdb4 on /media/sdb4 type ext3 (rw,relatime,sync,errors=continue,user_xattr,barrier=1,data=ordered)
/dev/sdb3 on /media/sdb3 type ext3 (rw,relatime,sync,errors=continue,user_xattr,barrier=1,data=ordered)
/dev/sda1 on /boot/efi type vfat (rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro)
/dev/sdb1 on /mnt type vfat (rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro)

# rm -f /sys/firmware/efi/efivars/BootNext* /sys/firmware/efi/efivars/fwupdate* /sys/firmware/efi/efivars/Boot0000*

# hexdump -C fwupdate-e0f614ed-fb82-467a-a34e-71172cc07e4d-0-0abba7dc-e516-4167-bbf5-4d9d1c739416
00000000  07 00 00 00 07 00 00 00  ed 14 f6 e0 82 fb 7a 46  |..............zF|
00000010  a3 4e 71 17 2c c0 7e 4d  00 00 00 00 00 00 00 00  |.Nq.,.~M........|
00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000030  00 00 00 00 01 00 00 00  02 01 0c 00 d0 41 03 0a  |.............A..|
00000040  00 00 00 00 01 01 06 00  00 13 03 12 0a 00 00 00  |................|
00000050  ff ff 00 00 04 01 2a 00  01 00 00 00 00 20 00 00  |......*...... ..|
00000060  00 00 00 00 00 20 00 00  00 00 00 00 6a a5 3a ce  |..... ......j.:.|
00000070  5b 51 37 49 a1 f2 b5 a1  a8 e6 f4 74 02 02 04 04  |[Q7I.......t....|
00000080  46 00 5c 00 45 00 46 00  49 00 5c 00 64 00 65 00  |F.\.E.F.I.\.d.e.|
00000090  6c 00 6c 00 5c 00 66 00  77 00 5c 00 66 00 77 00  |l.l.\.f.w.\.f.w.|
000000a0  75 00 70 00 64 00 61 00  74 00 65 00 2d 00 64 00  |u.p.d.a.t.e.-.d.|
000000b0  38 00 42 00 34 00 5a 00  63 00 2e 00 63 00 61 00  |8.B.4.Z.c...c.a.|
000000c0  70 00 00 00 7f ff 04 00                           |p.......|
000000c8

# # hexdump -C BootNext*
00000000  07 00 00 00 00 00                                 |......|
00000006

# hexdump -C Boot0000*
00000000  07 00 00 00 01 00 00 00  60 00 4c 00 69 00 6e 00  |........`.L.i.n.|
00000010  75 00 78 00 20 00 46 00  69 00 72 00 6d 00 77 00  |u.x. .F.i.r.m.w.|
00000020  61 00 72 00 65 00 20 00  55 00 70 00 64 00 61 00  |a.r.e. .U.p.d.a.|
00000030  74 00 65 00 72 00 00 00  04 01 2a 00 01 00 00 00  |t.e.r.....*.....|
00000040  00 20 00 00 00 00 00 00  00 20 00 00 00 00 00 00  |. ....... ......|
00000050  6a a5 3a ce 5b 51 37 49  a1 f2 b5 a1 a8 e6 f4 74  |j.:.[Q7I.......t|
00000060  02 02 04 04 32 00 5c 00  45 00 46 00 49 00 5c 00  |....2.\.E.F.I.\.|
00000070  64 00 65 00 6c 00 6c 00  5c 00 66 00 77 00 75 00  |d.e.l.l.\.f.w.u.|
00000080  70 00 64 00 61 00 74 00  65 00 2e 00 65 00 66 00  |p.d.a.t.e...e.f.|
00000090  69 00 00 00 7f ff 04 00                           |i.......|
00000098

# ls /dev/disk/by-uuid/ -l
total 0
lrwxrwxrwx 1 root root 10 Oct 15 02:55 1f66bb01-dfcb-4e84-abff-d873016bde4e -> ../../sdb4
lrwxrwxrwx 1 root root 10 Oct 15 02:55 2ef3b596-392d-43e2-9ca3-00fd747c14a3 -> ../../sdb2
lrwxrwxrwx 1 root root 10 Oct 15 02:56 561D-FFCC -> ../../sda1
lrwxrwxrwx 1 root root 10 Oct 15 02:55 AE67-7F11 -> ../../sdb1
lrwxrwxrwx 1 root root 10 Oct 15 02:55 f16d9979-60e3-48a6-abf5-d9505faf8fa1 -> ../../sdb3

# ls /dev/disk/
by-id  by-path  by-uuid

# ls -l /sys/dev/block/
total 0
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:0 -> ../../devices/virtual/block/ram0
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:1 -> ../../devices/virtual/block/ram1
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:10 -> ../../devices/virtual/block/ram10
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:11 -> ../../devices/virtual/block/ram11
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:12 -> ../../devices/virtual/block/ram12
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:13 -> ../../devices/virtual/block/ram13
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:14 -> ../../devices/virtual/block/ram14
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:15 -> ../../devices/virtual/block/ram15
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:2 -> ../../devices/virtual/block/ram2
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:3 -> ../../devices/virtual/block/ram3
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:4 -> ../../devices/virtual/block/ram4
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:5 -> ../../devices/virtual/block/ram5
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:6 -> ../../devices/virtual/block/ram6
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:7 -> ../../devices/virtual/block/ram7
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:8 -> ../../devices/virtual/block/ram8
lrwxrwxrwx 1 root root 0 Oct 15 02:55 1:9 -> ../../devices/virtual/block/ram9
lrwxrwxrwx 1 root root 0 Oct 15 02:55 7:0 -> ../../devices/virtual/block/loop0
lrwxrwxrwx 1 root root 0 Oct 15 02:55 7:1 -> ../../devices/virtual/block/loop1
lrwxrwxrwx 1 root root 0 Oct 15 02:55 7:2 -> ../../devices/virtual/block/loop2
lrwxrwxrwx 1 root root 0 Oct 15 02:55 7:3 -> ../../devices/virtual/block/loop3
lrwxrwxrwx 1 root root 0 Oct 15 02:55 7:4 -> ../../devices/virtual/block/loop4
lrwxrwxrwx 1 root root 0 Oct 15 02:55 7:5 -> ../../devices/virtual/block/loop5
lrwxrwxrwx 1 root root 0 Oct 15 02:55 7:6 -> ../../devices/virtual/block/loop6
lrwxrwxrwx 1 root root 0 Oct 15 02:55 7:7 -> ../../devices/virtual/block/loop7
lrwxrwxrwx 1 root root 0 Oct 15 02:55 8:0 -> ../../devices/pci0000:00/0000:00:13.0/ata1/host0/target0:0:0/0:0:0:0/block/sda
lrwxrwxrwx 1 root root 0 Oct 15 02:55 8:1 -> ../../devices/pci0000:00/0000:00:13.0/ata1/host0/target0:0:0/0:0:0:0/block/sda/sda1
lrwxrwxrwx 1 root root 0 Oct 15 02:55 8:16 -> ../../devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/host2/target2:0:0/2:0:0:0/block/sdb
lrwxrwxrwx 1 root root 0 Oct 15 02:55 8:17 -> ../../devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/host2/target2:0:0/2:0:0:0/block/sdb/sdb1
lrwxrwxrwx 1 root root 0 Oct 15 02:55 8:18 -> ../../devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/host2/target2:0:0/2:0:0:0/block/sdb/sdb2
lrwxrwxrwx 1 root root 0 Oct 15 02:55 8:19 -> ../../devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/host2/target2:0:0/2:0:0:0/block/sdb/sdb3
lrwxrwxrwx 1 root root 0 Oct 15 02:55 8:20 -> ../../devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/host2/target2:0:0/2:0:0:0/block/sdb/sdb4

# echo -e -n "\x07\x00\x00\x00\x01" > /sys/firmware/efi/efivars/SHIM_DEBUG-605dab50-e046-4300-abb6-3dd810dd8b23
```

After reboot, the fwupdate.efi is not launched at all. To work around this issue, I made such a patch and it works well:

```
 static EFI_STATUS
+search_file(EFI_DEVICE_PATH **file_dp, EFI_FILE_HANDLE *fh)
+{
+       EFI_DEVICE_PATH *dp, *parent_dp;
+       EFI_GUID sfsp = SIMPLE_FILE_SYSTEM_PROTOCOL;
+       EFI_GUID dpp = DEVICE_PATH_PROTOCOL;
+       EFI_HANDLE *devices;
+       UINTN i, n_handles, count;
+       EFI_STATUS rc;
+
+       rc = uefi_call_wrapper(BS->LocateHandleBuffer, 5, ByProtocol, &sfsp, NULL,
+                              &n_handles, &devices);
+       if (EFI_ERROR(rc)) {
+               Print(L"Could not find handles.\n");
+               return rc;
+       }
+
+       dp = *file_dp;
+
+       if (debugging)
+               Print(L"Searching Device Path: %s ...\n", DevicePathToStr(dp));
+
+       parent_dp = DuplicateDevicePath(dp);
+       if (!parent_dp) {
+               rc = EFI_INVALID_PARAMETER;
+               goto out;
+       }
+
+       dp = parent_dp;
+       count = 0;
+       while (1) {
+               if (IsDevicePathEnd(dp)) {
+                       rc = EFI_INVALID_PARAMETER;
+                       goto out;
+               }
+       
+               if (DevicePathType(dp) == MEDIA_DEVICE_PATH && DevicePathSubType(dp) == MEDIA_FILEPATH_DP)
+                       break;
+
+               dp = NextDevicePathNode(dp);
+               ++count;
+       }
+
+       SetDevicePathEndNode(dp);
+
+       if (debugging)
+               Print(L"Device Path prepared: %s\n", DevicePathToStr(parent_dp));
+
+       for (i = 0; i < n_handles; i++) {
+               EFI_DEVICE_PATH *path;
+
+       for (i = 0; i < n_handles; i++) {
+               EFI_DEVICE_PATH *path;
+
+               rc = uefi_call_wrapper(BS->HandleProtocol, 3, devices[i], &dpp,
+                                      (void **)&path);
+               if (EFI_ERROR(rc))
+                       continue;
+
+               if (debugging)
+                       Print(L"Device supporting SFSP: %s\n", DevicePathToStr(path));
+
+               rc = EFI_UNSUPPORTED;
+               while (!IsDevicePathEnd(path)) {
+                       if (debugging)
+                               Print(L"Comparing: %s and %s\n", DevicePathToStr(parent_dp), DevicePathToStr(path));
+
+                       if (LibMatchDevicePaths(path, parent_dp) == TRUE) {
+                               *fh = devices[i];
+                               for (i = 0; i < count; i++)
+                                       *file_dp = NextDevicePathNode(*file_dp);
+                               rc = EFI_SUCCESS;
+
+                               if (debugging)
+                                       Print(L"Match up! Returning %s\n", DevicePathToStr(*file_dp));
+
+                               goto out;
+                       }
+
+                       path = NextDevicePathNode(path);
+               }
+       }
+
+out:
+       if (!EFI_ERROR(rc))
+               Print(L"File %s searched\n", DevicePathToStr(*file_dp));
+
+       uefi_call_wrapper(BS->FreePool, 1, devices);
+       return rc;
+}
+
+static EFI_STATUS
 open_file(EFI_DEVICE_PATH *dp, EFI_FILE_HANDLE *fh)
 {
        EFI_DEVICE_PATH *file_dp = dp;
@@ -373,9 +461,12 @@ open_file(EFI_DEVICE_PATH *dp, EFI_FILE_HANDLE *fh)
        rc = uefi_call_wrapper(BS->LocateDevicePath, 3, &sfsp, &file_dp,
                               &device);
        if (EFI_ERROR(rc)) {
-               Print(L"%a:%a():%d: Could not locate device handle: %r\n",
-                             __FILE__, __func__, __LINE__, rc);
-               return rc;
+               rc = search_file(&file_dp, &device);
+               if (EFI_ERROR(rc)) {
+                       Print(L"%a:%a():%d: Could not locate device handle: %r\n",
+                                     __FILE__, __func__, __LINE__, rc);
+                       return rc;
+               }
        }

        if (DevicePathType(file_dp) != MEDIA_DEVICE_PATH ||
@@ -752,6 +753,11 @@ efi_main(EFI_HANDLE image, EFI_SYSTEM_TABLE *systab)
                return rc;
        }

+       if (debugging) {
+               Print(L"Applying the capsules ...\n");
+               uefi_call_wrapper(BS->Stall, 1, 10000000);
+       }
+
        /*
         * Step 4: apply the capsules.
         */
```

If fwupdate employs USB key to execute capsule update, the generated device path is valid. However, the fwupdate.efi will report open_file() fails. With the patch above, USB key also works well.

Here are the dumps for USB key as ESP

```
# hexdump -C fwupdate-e0f614ed-fb82-467a-a34e-71172cc07e4d-0-0abba7dc-e516-4167-bbf5-4d9d1c739416 
00000000  07 00 00 00 07 00 00 00  ed 14 f6 e0 82 fb 7a 46  |..............zF|
00000010  a3 4e 71 17 2c c0 7e 4d  00 00 00 00 00 00 00 00  |.Nq.,.~M........|
00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000030  00 00 00 00 01 00 00 00  04 01 2a 00 01 00 00 00  |..........*.....|
00000040  00 08 00 00 00 00 00 00  00 00 04 00 00 00 00 00  |................|
00000050  89 8c c8 59 00 00 00 00  00 00 00 00 00 00 00 00  |...Y............|
00000060  01 01 04 04 46 00 5c 00  45 00 46 00 49 00 5c 00  |....F.\.E.F.I.\.|
00000070  64 00 65 00 6c 00 6c 00  5c 00 66 00 77 00 5c 00  |d.e.l.l.\.f.w.\.|
00000080  66 00 77 00 75 00 70 00  64 00 61 00 74 00 65 00  |f.w.u.p.d.a.t.e.|
00000090  2d 00 37 00 45 00 42 00  36 00 6c 00 31 00 2e 00  |-.7.E.B.6.l.1...|
000000a0  63 00 61 00 70 00 00 00  7f ff 04 00              |c.a.p.......|
000000ac
# hexdump -C BootNext-8be4df61-93ca-11d2-aa0d-00e098032b8c 
00000000  07 00 00 00 00 00                                 |......|
00000006
# hexdump -C Boot0000-8be4df61-93ca-11d2-aa0d-00e098032b8c 
00000000  07 00 00 00 01 00 00 00  60 00 4c 00 69 00 6e 00  |........`.L.i.n.|
00000010  75 00 78 00 20 00 46 00  69 00 72 00 6d 00 77 00  |u.x. .F.i.r.m.w.|
00000020  61 00 72 00 65 00 20 00  55 00 70 00 64 00 61 00  |a.r.e. .U.p.d.a.|
00000030  74 00 65 00 72 00 00 00  04 01 2a 00 01 00 00 00  |t.e.r.....*.....|
00000040  00 08 00 00 00 00 00 00  00 00 04 00 00 00 00 00  |................|
00000050  89 8c c8 59 00 00 00 00  00 00 00 00 00 00 00 00  |...Y............|
00000060  01 01 04 04 32 00 5c 00  45 00 46 00 49 00 5c 00  |....2.\.E.F.I.\.|
00000070  64 00 65 00 6c 00 6c 00  5c 00 66 00 77 00 75 00  |d.e.l.l.\.f.w.u.|
00000080  70 00 64 00 61 00 74 00  65 00 2e 00 65 00 66 00  |p.d.a.t.e...e.f.|
00000090  69 00 00 00 7f ff 04 00                           |i.......|
00000098
```

The picture attached shows how the patch works (note that the picture is captures recently so it doesn't match the name of capsule file shown in the dumps):

![usb](https://cloud.githubusercontent.com/assets/10241344/10504020/8974c194-732e-11e5-8624-e0d90bf80f3c.png)

**Comment by jiazhang0** [comment]
After a re-creation of GPT on SSD, the first issue is gone. Maybe the relict of GPT is wrongly scanned by efivar and then reports to fwupdate.

**Comment by jiazhang0** [comment]
This issue should be closed and move to efivar. The root cause is that efivar will try to search the backup GPT Partition Entry Array and header to get the UUID of partition even they are already corrupted. For example, the end user runs parted to create a ESP with GPT scheme but not rm it. Then running dd to clean the primary GPT Partition Entry Array and header. At this point, Linux kernel cannot recognize current disk with valid GPT scheme and thus cannot find out the ESP on disk.  efivar can still find the UUID of partition from the backup GPT Partition Entry Array and thus generate the wrong device path for status variable and Boot####. efivar should add the sanity check to confirm whether or not a valid GPT exists.

**Comment by superm1** [comment]
@jiazhang0 Does this mean that the issue can also be avoided by running partprobe to make the kernel re-read partition tables after dd'ing the image?

**Comment by jiazhang0** [comment]
@superm1 partprobe doesn't halp. efivar tries the best to search primary and backup GPT partition entry array to find the valid partition entry info. Actually, the protective MBR has been corrputed. efivar should add a check to make sure the protective MBR is valid then make furthter check to validate GPT.

**Comment by jiazhang0** [comment]
Reopen this issue because we have a fix now for fwupdate.  See https://github.com/rhinstaller/fwupdate/pull/23

**Comment by superm1** [comment]
@jiazhang0 can't this be closed since #23 was closed then?  Or do you think there should be a second fix for this?

**Comment by jiazhang0** [comment]
@superm1 Yes. Actually there is another issue as I mentioned initially (Need a patch applied to fwupdate.efi). I will submit another fix later. Close this issue.
